:sparkles: driver / http / add context (#266)
alakarteio/k-ramel: packages/drivers/http/README.md

 - Main repository: [k-ramel](https://github.com/alakarteio/k-ramel)

# Installation
This driver is added by default to k-ramel stores.

# How to use it
In a reaction (See main documentation about listeners/reactions)

```js
export const login = (store, action, drivers) => {
  const { http } = drivers
  const todo = store.data.todos.get(action.payload)

  // TODOS is the context for this request
  // You can then catch events like `/@@http\/TODOS>.*/
  // You have the possibility to add more context with an object on second parameters
  http('TODOS', { myContext: 'todo one' })
    // It will
    // 1. trigger an event like `@@http/TODOS>POST>STARTED`
    // 2. serialize your todo
    // 3. add the header Content-Type to application/json
    // 4. use fetch to call your API
    // 5. trigger event like `@@http/TODOS>POST>FAILED` or `@@http/TODOS>POST>ENDED`
    // 6. if you have added a `context`, it is return on event ;)
    .post('/api/todos', todo)
}
```

# Emitted events
| event type | when |
|---|---|

> |`@@http/MY_CONTEXT>POST>STARTED`| when your fetch is started |

Reviewer: ```suggestion
|`@@http/MY_CONTEXT>POST>STARTED`| dispatched just before the fetch is started |
```

> |`@@http/MY_CONTEXT>POST>ENDED`| when your fetch is finished with status "OK" and without exception |

Reviewer: ```suggestion
|`@@http/MY_CONTEXT>POST>ENDED`| dispatched once the request is done without error (based on HTTP status) |
```

> |`@@http/MY_CONTEXT>POST>FAILED`| when your fetch respond with errors |

Reviewer: ```suggestion
|`@@http/MY_CONTEXT>POST>FAILED`| dispatched once the request is done **with** error (based on HTTP status) | |
```

> # Event data

Reviewer: ```suggestion
# Actions
```

All emitted events have some data

> | data | description | type |

Reviewer: ```suggestion
| field | description | type |
```

|---|---|---|
| `type` | event type | `string` | 

> | `fetch` | url and fetch options | `array` (`[url, options]`) |

Reviewer: ```suggestion
| `fetch` | url and fetch options, this can be used to retry a request that is on error | `array` (`[url, options]`) |
```

> | `status` | status of respond | `number` |

Reviewer: ```suggestion
| `status` | HTTP status of the response | `number` |
```

> | `payload` | result data of respond | `any` |

Reviewer: ```suggestion
| `payload` | data of the response (if the content type is `application/json`, the data is already parsed) | `any` |
```

> | `context` | your optional context added when use the driver | `any` |

Reviewer: See previous comment -> explain WHY we added this

  // TODOS is the context for this request
  // You can then catch events like `/@@http\/TODOS>.*/

>   // You have the possibility to add more context with an object on second parameters

Reviewer: I prefer that you add a new paragraph and explain WHY we should (or not) context object

Reviewer: And how
